ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,884 @@
+      LogSegment logSegment = srcLog.getSegment(logSegmentsUnderCompaction.get(0));
+      LogSegment prevSegment = srcLog.getPrevSegment(logSegment);
+      if (prevSegment != null) {
+        // duplicate data, if it exists, can only be in the previous log segment.
+        Offset startOffset = new Offset(prevSegment.getName(), prevSegment.getStartOffset());
+        Offset endOffset = new Offset(prevSegment.getName(), prevSegment.getEndOffset());
+        duplicateSearchSpan = new FileSpan(startOffset, endOffset);
+      }
+    }
+
+    for (String logSegmentName : logSegmentsUnderCompaction) {
+      LogSegment srcLogSegment = srcLog.getSegment(logSegmentName);
+      Offset logSegmentEndOffset = new Offset(srcLogSegment.getName(), srcLogSegment.getEndOffset());
+      if (needsCopying(logSegmentEndOffset) && !copyDataByLogSegment(srcLogSegment, duplicateSearchSpan,
+          startOffsetOfLastIndexSegmentForDelete)) {
+        if (isActive) {
+          compactionLog.splitCurrentCycle(logSegmentName);
+        }
+        // else shutdown is in progress.
+        break;
+      }
+      // only the very first log segment in the cycle can contain duplicate records.
+      duplicateSearchSpan = null;
+    }
+
+    numSwapsUsed = tgtIndex.getLogSegmentCount();
+    tgtIndex.close();
+    tgtLog.close();
+    // persist the bloom of the "latest" index segment if it exists
+    if (numSwapsUsed > 0) {
+      tgtIndex.getIndexSegments().lastEntry().getValue().map(true);
+    } else {
+      // there were no valid entries copied, return any temp segments back to the pool
+      File[] files = dataDir.listFiles(TEMP_LOG_SEGMENTS_FILTER);
+      if (files != null) {
+        for (File file : files) {
+          // TODO (DiskManager changes): This will actually return the segment to the DiskManager pool.
+          if (!file.delete()) {
+            throw new IllegalStateException("Could not delete segment file: " + file.getAbsolutePath());
+          }
+        }
+      } else {
+        throw new StoreException("Could not list temp files in directory: " + dataDir, StoreErrorCodes.Unknown_Error);
+      }
+    }
+  }
+
+  /**
+   * Commits the changes made by copying valid data. Commit involves
+   * 1. Renaming temporary log segments
+   * 2. Adding them to the log segments maintained by the application log.
+   * 3. Atomically switching the old set of index segments for the new ones (if not recovering).
+   * @param recovering {@code true} if this function was called in the context of recovery. {@code false} otherwise.
+   * @throws IOException if there were I/O errors during committing.
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  private void commit(boolean recovering) throws IOException, StoreException {
+    List<String> logSegmentNames = getTargetLogSegmentNames();
+    renameLogSegments(logSegmentNames);
```

Comment:
Wanted to avoid successive changes to the `validIndexSegments` in `PersistentIndex`. While there are no correctness issues, I would rather batch it. I don't see a huge downside with this approach.